### PR TITLE
Relative permeability

### LINKED ISSUE #27: Relative permeability
Container for analyzing and visualizing relative permeability models used in reservoir simulations.

### PR DESCRIPTION
**Pull request by asnyv**
Resolves #27

### CODE REVIEW

File: .github/workflows/webviz-subsurface.yml [hunk 1]
```diff
@@ -52,17 +55,35 @@ jobs:
...
-        TESTDATA_REPO_BRANCH: master
+        TESTDATA_REPO_BRANCH: relperm
+
       run: |
         webviz certificate
         git clone --depth 1 --branch $TESTDATA_REPO_BRANCH https://github.com/$TESTDATA_REPO_OWNER/webviz-subsurface-testdata.git
         pip install git+https://github.com/equinor/fmu-ensemble
+        sudo apt-get update
+        sudo apt-get install libboost-all-dev
+        sudo apt-get install liblapack-dev
```
Comment (anders-kiaer) [inline .github/workflows/webviz-subsurface.yml]:
```suggestion
        sudo apt-get install libboost-all-dev liblapack-dev
```

File: setup.py [hunk 2]
```diff
@@ -49,6 +50,7 @@
         "pandas~=0.24",
         "pillow~=6.1",
         "xtgeo~=2.1",
+        "pyscal~=0.4.1",
```
Comment (anders-kiaer) [inline setup.py]:
```suggestion
        "pyscal~=0.4, >=0.4.1",
```

File: webviz_subsurface/_datainput/relative_permeability.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return load_ensemble_set(ensemble_paths, ensemble_set_name).apply(ecl2df_satfunc)
+
+
+@CACHE.memoize(timeout=CACHE.TIMEOUT)
+@webvizstore
+def load_scal_recommendation(
+    scalfile: Path, sheet_name: Optional[Union[str, int, list]] = None
+) -> pd.DataFrame:
+    scalfile = str(scalfile)  # pyscal currently can't handle a pathlib.Path
+    scal_data = PyscalFactory.load_relperm_df(scalfile, sheet_name)
```
Comment (anders-kiaer) [inline webviz_subsurface/_datainput/relative_permeability.py]:
```suggestion
    scal_data = PyscalFactory.load_relperm_df(str(scalfile), sheet_name)
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+            self.family = 1
+            if any(
+                keyword in RelativePermeability.RELPERM_FAMILIES[2]
+                for keyword in self.satfunc["KEYWORD"].unique()
+            ):
+                warnings.warn(
+                    (
+                        f"Mix of keyword family 1 and 2, currently only support one family at the "
+                        f"time. Dropping all data of family 2 ('SWFN', 'SGFN', 'SGWFN', 'SOF2', "
+                        f"'SOF3', 'SOF32D') and continues with family 1 ('SWOF', 'SGOF', 'SLGOF')."
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
                        "Mix of keyword family 1 and 2, currently only support one family at the "
                        "time. Dropping all data of family 2 ('SWFN', 'SGFN', 'SGWFN', 'SOF2', "
                        "'SOF3', 'SOF32D') and continues with family 1 ('SWOF', 'SGOF', 'SLGOF')."
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+                        RelativePermeability.RELPERM_FAMILIES["fam1"]
+                    )
+                ]
+            if "SGOF" in self.satfunc["KEYWORD"].unique():
+                if "SLGOF" in self.satfunc["KEYWORD"].unique():
+                    warnings.warn(
+                        (
+                            f"Mix of 'SGOF' and 'SLGOF' in ensembles, resulting in non-unique "
+                            f"horizontal axis ('SG' and 'SL') for 'KRG', 'KROG' and 'PCOG'. "
+                            f"Dropping all data with 'SLGOF'."
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
                            "Mix of 'SGOF' and 'SLGOF' in ensembles, resulting in non-unique "
                            "horizontal axis ('SG' and 'SL') for 'KRG', 'KROG' and 'PCOG'. "
                            "Dropping all data with 'SLGOF'."
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+                    "SW": ["KRW", "KROW", "PCOW"],
+                    "SL": ["KRG", "KROG", "PCOG"],
+                }
+        elif not all(
+            keyword in RelativePermeability.RELPERM_FAMILIES[2]
+            for keyword in self.satfunc["KEYWORD"].unique()
+        ):
+            raise ValueError(
+                f"Unrecognized saturation table keyword in data. This should not occur unless "
+                f"there has been changes to ecl2df. Update of this plugin might be required."
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
                "Unrecognized saturation table keyword in data. This should not occur unless "
                "there has been changes to ecl2df. Update of this plugin might be required."
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+                "SO": ["KROW", "KROG"],
+            }
+        self.scalfile = scalfile
+        self.sheet_name = sheet_name
+        self.scal = (
+            load_scal_recommendation(self.scalfile, self.sheet_name)
+            if self.scalfile is not None
+            else None
+        )
+        self.plot_options = options if options is not None else {}
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
        self.plot_options = {} if options is None else options
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+        self.set_callbacks(app)
+
+    @property
+    # pylint: disable=consider-iterating-dictionary
+    def sat_axes(self):
+        """List of all possible saturation axes in dataframe"""
+        return [sat for sat in self.sat_axes_maps.keys() if sat in self.satfunc.columns]
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
    def sat_axes(self):
        """List of all possible saturation axes in dataframe"""
        return [sat for sat in self.sat_axes_maps if sat in self.satfunc.columns]
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+                stored_satnum.get("SATNUM", self.satnums[0]),
+            )
+
+        @app.callback(
+            [
+                Output(self.uuid("curve"), "value"),
+                Output(self.uuid("curve"), "options"),
+            ],
+            [Input(self.uuid("sataxis"), "value")],
+            # [State(self.uuid("stored_curves"), "data")],
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
```

File: webviz_subsurface/plugins/_relative_permeability.py [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+                load_satfunc,
+                [
+                    {
+                        "ensemble_paths": self.ens_paths,
+                        "ensemble_set_name": "EnsembleSet",
+                    }
+                ],
+            )
+        )
+        return functions
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_relative_permeability.py]:
```suggestion
    return [(load_satfunc, [{"ensemble_paths": self.ens_paths, "ensemble_set_name": "EnsembleSet"}])] +
    [] if self.scalfile is None else [(load_scal_recommendation, [{"scalfile": self.scalfile, "sheet_name": self.sheet_name}])]
```

File: webviz_subsurface/plugins/_relative_permeability.py (hunk 9 above)
Comment (asnyv) [reply webviz_subsurface/plugins/_relative_permeability.py:513]: removed

**Pull request merged**